README.md
Infinite List
===========

Infinite List

## About

[Infinite List GitHub Pages](http://rapid-application-development-js.github.io/InfiniteList/)

[RAD.JS: Tookit & framework for javascript application development](https://github.com/Rapid-Application-Development-JS/)

[RAD.JS Blog](http://rad-js.com/)

[MobiDev Software Development Company](http://mobidev.biz/)

## Advantages:

* Lightweight;
* Cross browser compatible;
* Tested on different devices and platforms;
* High performance, on the par and faster than other libraries;

## Install

Install using Bower.

```bash
bower install --save radjs-infinitelist
```

Or using npm.

```bash
npm install --save radjs-infinitelist
```


doc/NEWS.md
Infinite List Release Notes
==========================

Version 1.0.0 - Jun 18, 2015
---------------------

Initial release.


gulpfile.js
var gulp = (function () {
	var config = require('./config.json');
	var libName = config['name'];
	var sourceFiles = config['sourceFiles'];
	var sourceWebComponent = config['sourceWebComponent'];
	var folderSource = config['folderSource'];
	var folderCompiled = config['folderCompiled'];
	var gulp = require('gulp');
	var gulpConcat = require('gulp-concat');
	var gulpSourceMaps = require('gulp-sourcemaps');
	var gulpRename = require('gulp-rename');
	var gulpUglify = require('gulp-uglify');
	gulp.task('scripts', function () {
		return gulp.src(sourceFiles)
			.pipe(gulpConcat(libName + '.js'))
			.pipe(gulp.dest('./' + folderCompiled));
	});
	gulp.task('watch', function () {
		gulp.watch('./' + folderSource, ['default']);
	});
	gulp.task('default', ['scripts'], function () {
		gulp
			.src([
				folderCompiled + '/' + libName + '.js'
				, folderCompiled + '/x-' + libName + '.js'
			])
			.pipe(gulpRename({
				extname: '.min.js'
			}))
			.pipe(gulpSourceMaps.init())
			.pipe(gulpUglify({
				compress: {
					sequences: false,
					unsafe: false,
					warnings: true
				}
			}))
			.pipe(gulpSourceMaps.write('.'))
			.pipe(gulp.dest(folderCompiled));
	});
	return gulp;
})();


example/js/index.js
(function () {
	/**
	 * Helper to generate colors
	 * @type {{_colors: string[], random: Function}}
	 */
	var ColorGen = {
		_colors: [
			"0,153,229", "0,167,142", "2,100,102", "26,214,253", "52,170,220",
			"52,191,73", "82,237,199", "90,212,39", "125,63,152", "135,252,112",
			"155,89,182", "198,68,252", "200,110,223", "207,220,0", "239,77,182",
			"243,236,25", "255,73,129", "255,76,76", "255,94,58", "255,149,0",
			"255,217,0", "55,219,76"
		],
		random: function (opacity) {
			return "rgba(" + ColorGen._colors[Math.floor(Math.random() * ColorGen._colors.length)] + "," + opacity + ")";
		}
	};
	/**
	 * Adapter object
	 * @param {String} direction
	 * @param {Function} onElementCreated
	 * @constructor
	 */
	function Adapter(direction, onElementCreated) {
		this.direction = direction;
		this.onElementCreated = onElementCreated.bind(this) || new Function;
	}
	Adapter.prototype = {
		/**
		 * Get total count of items
		 * @return {number}
		 * @public
		 */
		getCount: function () {
			return 1000;
		},
		/**
		 * Get item as HTMLElement
		 * @param {Number} index
		 * @param {Object} element
		 * @return {HTMLElement|Object}
		 * @public
		 */
		getItem: function (index, element) {
			// Create item element if it doesn"t exist
			if (!element) {
				element = document.createElement("div");
				this.onElementCreated(element);
			}
			// Fill by content
			element.innerHTML = "<div>" + index + "</div>";
			return element;
		}
	};
	// Application initizlization
	var direction = ["vertical", "horizontal"][0];
	var container = document.querySelector(".container");
	container.classList.add(direction);
	var infiniteListStyling = function (element) {
		element.className = "item";
		element.classList.add("item");
		element.classList.add(this.direction);
		element.style.backgroundColor = ColorGen.random(.75);
	};
	var adapter = new Adapter(direction, infiniteListStyling);
	container.list = new RADJS_InfiniteList(container, adapter, direction);
}());


source/infinitelist.js
/**
 * @param {Number} index
 * @param {Element} item
 * @constructor
 */
function ItemHolder (index, item) {
	this.index = index;
	this.item = item;
	ItemHolder.prototype = {
		index: -1,
		item: undefined,
		position: 0,
		visible: true
	};
}
/**
 * @param {HTMLElement} container
 * @param {Object} adapter
 * @param {String} direction
 * @constructor
 */
function InfiniteList(container, adapter, direction) {
	if (container instanceof Element === false) {
		throw "Container is not an HTML element";
	}
	if (container.firstElementChild === null) {
		throw "Object's first child element is empty";
	}
	if (!adapter || typeof adapter.getCount !== "function" || typeof adapter.getItem !== "function") {
		throw "Assigned adapter has no support of `getCount` and `getItem` methods";
	}
	direction = (direction + "").toLowerCase();
	if (["h", "v", "horizontal", "vertical"].indexOf(direction) === -1) {
		throw "Wrong scroll direction [" + direction + "]";
	}
	// Assign arguments
	this.adapter = adapter; // {getItem, getItem}
	this.elContainer = container; // HTML Element
	this.elWrapper = container.firstElementChild; // HTML Element
	this.isVertical = direction === "vertical" || direction === "v";
	// Construction part
	this.items = []; // Items in scroll
	this.itemsInvisible = []; // Invisile items
	this.requestAnimationFrameStep = this.requestAnimationFrameStep.bind(this);
	this.onScroll = this.onScroll.bind(this);
	this.refresh = this.refresh.bind(this);
	container.addEventListener("scroll", this.onScroll, false);
	window.addEventListener("resize", this.refresh, false);
	// Create destroy method
	this.destroy = function () {
		container.removeEventListener("scroll", this.onScroll);
		window.removeEventListener("resize", this.refresh);
		this.state.inAnimation = false;
		window.cancelAnimationFrame(this.requestAnimationFrameID);
		this.elWrapper.innerHTML = "";
		this.adapter = null;
		this.elContainer = null;
		this.elWrapper = null;
		this.items = null;
		this.itemsInvisible = null;
		delete this.requestAnimationFrameStep;
		delete this.onScroll;
		delete this.refresh;
	};
	// Update full status
	this.state = {};
	/**
	 * Is animation happening
	 * @type {boolean}
	 */
	this.state.inAnimation = false;
	/**
	 * Heigth or width of item in pixels
	 * @type {Number}
	 */
	this.state.itemSize = 0;
	/**
	 * Height or width of container (Window)
	 * @type {Number}
	 */
	this.state.containerSize = 0;
	/**
	 * Index of element in bottom / right
	 * @type {Number}
	 */
	this.state.lastIndex = -1;
	/**
	 * Top padding from container
	 * @type {Number}
	 */
	this.state.containerPadding = 0;
	this.refresh(true);
}
InfiniteList.prototype = {
	isVertical: undefined,
	onScroll: function () {
		// Stop filling list
		window.cancelAnimationFrame(this.requestAnimationFrameID);
		this.state.inAnimation = false;
		// Give it rest
		this.requestAnimationFrameID = window.requestAnimationFrame(this.requestAnimationFrameStep);
	},
	requestAnimationFrameStep: function () {
		var elementPadCoord = propGet(this.elWrapper.getBoundingClientRect(), this.isVertical ? "top" : "left");
		if (this.state.containerPadding !== elementPadCoord) {
			this.fillList(this.state.containerPadding, elementPadCoord);
			this.state.containerPadding = elementPadCoord;
		}
		// Start new animation iteration
		if (this.state.inAnimation) {
			this.requestAnimationFrameID = window.requestAnimationFrame(this.requestAnimationFrameStep);
		}
	},
	/**
	 * Fill container list with items
	 * @param {Number} lastPos_elPadCoord
	 * @param {Number} newPosition_elPadCoord
	 */
	fillList: function (lastPos_elPadCoord, newPosition_elPadCoord) {
		var index, itemsLength, itemHolder, itemsCount = this.adapter.getCount(), minVisibleIndex, maxVisibleIndex;
		// Mark items as `invisible`
		for (index = 0, itemsLength = this.items.length; index < itemsLength; index += 1) {
			itemHolder = this.items[index];
			if (itemHolder.visible) {
				// Mark on top
				if ((itemHolder.position + this.state.itemSize + newPosition_elPadCoord) < 0) {
					itemHolder.visible = false;
					this.itemsInvisible.push(itemHolder);
				}
				// Mark on bottom
				if ((itemHolder.position + newPosition_elPadCoord) > this.state.containerSize) {
					itemHolder.visible = false;
					this.itemsInvisible.push(itemHolder);
					if (this.state.lastIndex >= itemHolder.index) {
						this.state.lastIndex = itemHolder.index - 1;
					}
				}
			}
		}
		// Find min and max real visible index
		for (index = 0, itemsLength = this.items.length; index < itemsLength; index += 1) {
			itemHolder = this.items[index];
			if (itemHolder.visible) {
				if (minVisibleIndex === undefined || itemHolder.index < minVisibleIndex) {
					minVisibleIndex = itemHolder.index;
				}
				if (maxVisibleIndex === undefined || itemHolder.index > maxVisibleIndex) {
					maxVisibleIndex = itemHolder.index;
				}
			}
		}
		// If we don't have visible items
		if (minVisibleIndex === undefined && maxVisibleIndex === undefined) {
			// Create fake bottom index
			this.state.lastIndex =
				Math.min(Math.floor(Math.abs(-newPosition_elPadCoord / this.state.itemSize)), itemsCount) || -1;
		}
		// Fill bottom part by new items
		while (
		(this.state.lastIndex < itemsCount - 1)
		&&
		(this.state.lastIndex * this.state.itemSize + newPosition_elPadCoord < this.state.containerSize)
			) {
			this.state.lastIndex += 1;
			this.applyItem(this.state.lastIndex);
		}
		// Fill top part with new items
		if (typeof minVisibleIndex === "number") {
			while (minVisibleIndex > 0 && minVisibleIndex * this.state.itemSize + newPosition_elPadCoord > 0) {
				minVisibleIndex -= 1;
				this.applyItem(minVisibleIndex);
			}
		}
		// Hide invisible items
		this.itemsInvisible.forEach(this.hideItem, this);
	},
	/**
	 * Hide item
	 * @param {Object} itemHolder
	 */
	hideItem: function (itemHolder) {
		propSet(itemHolder.item.style, this.isVertical ? "left" : "top" ,"200%");
	},
	/**
	 * Get height or width of item
	 * @return {Number}
	 */
	getItemSize: function () {
		var size, item;
		item = this.adapter.getItem(0, null);
		this.elWrapper.appendChild(item);
		size = propGet(item.getBoundingClientRect(), this.isVertical ? "height": "width");
		this.elWrapper.removeChild(item);
		return size;
	},
	/**
	 * Set item position in container
	 * @param {Object} itemHolder
	 */
	setItemPosition: function (itemHolder) {
		propSet(itemHolder.item.style, this.isVertical ? "top" : "left", itemHolder.position + "px");
		propSet(itemHolder.item.style, this.isVertical ? "left" : "top" , 0);
	},
	/**
	 * Create item holder if not exists
	 * @param {Number} index
	 * @return {Object}
	 */
	createNewItemHolder: function (index) {
		var item = this.adapter.getItem(index, null), itemHolder;
		this.elWrapper.appendChild(item);
		itemHolder = new ItemHolder(index, item);
		this.items.push(itemHolder);
		return itemHolder;
	},
	/**
	 * Apply item to container
	 * @param {Number} index
	 */
	applyItem: function (index) {
		var itemHolder = this.itemsInvisible.pop() || this.createNewItemHolder(index);
		itemHolder.position = index * this.state.itemSize;
		itemHolder.index = index;
		itemHolder.visible = true;
		this.adapter.getItem(itemHolder.index, itemHolder.item);
		this.setItemPosition(itemHolder);
	},
	/**
	 * Refresh scroll
	 * @param {boolean=true} noDelay
	 */
	refresh: function (noDelay) {
		var delay = noDelay ? 0 : 150;
		clearTimeout(this.refreshID);
		this.refreshID = setTimeout(function () {
			// Stop filling
			window.cancelAnimationFrame(this.requestAnimationFrameID);
			this.state.inAnimation = false;
			// Update state
			this.state.containerPadding =
				propGet(this.elWrapper.getBoundingClientRect(), this.isVertical ? "top" : "left");
			this.state.itemSize = this.getItemSize();
			this.state.containerSize = this.elContainer.offsetHeight;
			// Update wrapper size
			propSet(this.elWrapper.style, this.isVertical ? "height" : "width",
				this.state.itemSize * this.adapter.getCount() + "px");
			// Recalculate items positions
			for (var index = 0, itemLength = this.items.length; index < itemLength; index += 1) {
				this.items[index].position = this.items[index].index * this.state.itemSize;
			}
			// Fill list
			this.fillList(this.state.containerPadding, this.state.containerPadding);
		}.bind(this), delay);
	}
};


release/radjs-infinitelist.js
(function (root, factory) {
	if (typeof define === "function" && define.amd) {
		define("radjs-infinitelist", function () {
			root.RADJS_InfiniteList = factory();
			return root.RADJS_InfiniteList;
		});
	} else if (typeof module === "object" && module.exports) {
		root.RADJS_InfiniteList = factory();
		module.exports = root.RADJS_InfiniteList;
	} else {
		root.RADJS_InfiniteList = factory();
	}
}(this, function () {
	"use strict";

function propGet(target, key, empty) {
	return key in target ? target[key] : empty;
}

function propSet(target, key, value) {
	target[key] = value;
}

(function (_window) {
	_window.Date.now || (_window.Date.now = function () {
		return +new Date;
	});
})(window);

(function (_window) {
	_window.performance || (_window.performance = {});
	_window.performance.now && (_window.performance.now = performance.webkitNow || performance.mozNow || performance.msNow ||
		performance.oNow || function () {
			return Date.now()
		})
})(window);

(function (_window) {
	var lastTime = 0;
	var vendors = ["webkit", "moz", "ms", "o"];
	for (var x = 0; x < vendors.length && !_window.requestAnimationFrame; ++x) {
		_window.requestAnimationFrame = _window[vendors[x] + "RequestAnimationFrame"];
		_window.cancelAnimationFrame = _window[vendors[x] + "CancelAnimationFrame"]
			|| _window[vendors[x] + "CancelRequestAnimationFrame"];
	}
	_window.requestAnimationFrame || (_window.requestAnimationFrame = function (callback) {
		var currTime = Date.now();
		var timeToCall = Math.max(0, 16 - (currTime - lastTime));
		var id = _window.setTimeout(function () {
				callback(currTime + timeToCall)
			},
			timeToCall);
		lastTime = currTime + timeToCall;
		return id
	});
	_window.cancelAnimationFrame || (_window.cancelAnimationFrame = function (id) {
		clearTimeout(id);
	});
}(window));

/**
 * @param {Number} index
 * @param {Element} item
 * @constructor
 */
function ItemHolder (index, item) {
	this.index = index;
	this.item = item;
	ItemHolder.prototype = {
		index: -1,
		item: undefined,
		position: 0,
		visible: true
	};
}
/**
 * @param {HTMLElement} container
 * @param {Object} adapter
 * @param {String} direction
 * @constructor
 */
function InfiniteList(container, adapter, direction) {
	if (container instanceof Element === false) {
		throw "Container is not an HTML element";
	}
	if (container.firstElementChild === null) {
		throw "Object's first child element is empty";
	}
	if (!adapter || typeof adapter.getCount !== "function" || typeof adapter.getItem !== "function") {
		throw "Assigned adapter has no support of `getCount` and `getItem` methods";
	}
	direction = (direction + "").toLowerCase();
	if (["h", "v", "horizontal", "vertical"].indexOf(direction) === -1) {
		throw "Wrong scroll direction [" + direction + "]";
	}
	// Assign arguments
	this.adapter = adapter; // {getItem, getItem}
	this.elContainer = container; // HTML Element
	this.elWrapper = container.firstElementChild; // HTML Element
	this.isVertical = direction === "vertical" || direction === "v";
	// Construction part
	this.items = []; // Items in scroll
	this.itemsInvisible = []; // Invisile items
	this.requestAnimationFrameStep = this.requestAnimationFrameStep.bind(this);
	this.onScroll = this.onScroll.bind(this);
	this.refresh = this.refresh.bind(this);
	container.addEventListener("scroll", this.onScroll, false);
	window.addEventListener("resize", this.refresh, false);
	// Create destroy method
	this.destroy = function () {
		container.removeEventListener("scroll", this.onScroll);
		window.removeEventListener("resize", this.refresh);
		this.state.inAnimation = false;
		window.cancelAnimationFrame(this.requestAnimationFrameID);
		this.elWrapper.innerHTML = "";
		this.adapter = null;
		this.elContainer = null;
		this.elWrapper = null;
		this.items = null;
		this.itemsInvisible = null;
		delete this.requestAnimationFrameStep;
		delete this.onScroll;
		delete this.refresh;
	};
	// Update full status
	this.state = {};
	/**
	 * Is animation happening
	 * @type {boolean}
	 */
	this.state.inAnimation = false;
	/**
	 * Heigth or width of item in pixels
	 * @type {Number}
	 */
	this.state.itemSize = 0;
	/**
	 * Height or width of container (Window)
	 * @type {Number}
	 */
	this.state.containerSize = 0;
	/**
	 * Index of element in bottom / right
	 * @type {Number}
	 */
	this.state.lastIndex = -1;
	/**
	 * Top padding from container
	 * @type {Number}
	 */
	this.state.containerPadding = 0;
	this.refresh(true);
}
InfiniteList.prototype = {
	isVertical: undefined,
	onScroll: function () {
		// Stop filling list
		window.cancelAnimationFrame(this.requestAnimationFrameID);
		this.state.inAnimation = false;
		// Give it rest
		this.requestAnimationFrameID = window.requestAnimationFrame(this.requestAnimationFrameStep);
	},
	requestAnimationFrameStep: function () {
		var elementPadCoord = propGet(this.elWrapper.getBoundingClientRect(), this.isVertical ? "top" : "left");
		if (this.state.containerPadding !== elementPadCoord) {
			this.fillList(this.state.containerPadding, elementPadCoord);
			this.state.containerPadding = elementPadCoord;
		}
		// Start new animation iteration
		if (this.state.inAnimation) {
			this.requestAnimationFrameID = window.requestAnimationFrame(this.requestAnimationFrameStep);
		}
	},
	/**
	 * Fill container list with items
	 * @param {Number} lastPos_elPadCoord
	 * @param {Number} newPosition_elPadCoord
	 */
	fillList: function (lastPos_elPadCoord, newPosition_elPadCoord) {
		var index, itemsLength, itemHolder, itemsCount = this.adapter.getCount(), minVisibleIndex, maxVisibleIndex;
		// Mark items as `invisible`
		for (index = 0, itemsLength = this.items.length; index < itemsLength; index += 1) {
			itemHolder = this.items[index];
			if (itemHolder.visible) {
				// Mark on top
				if ((itemHolder.position + this.state.itemSize + newPosition_elPadCoord) < 0) {
					itemHolder.visible = false;
					this.itemsInvisible.push(itemHolder);
				}
				// Mark on bottom
				if ((itemHolder.position + newPosition_elPadCoord) > this.state.containerSize) {
					itemHolder.visible = false;
					this.itemsInvisible.push(itemHolder);
					if (this.state.lastIndex >= itemHolder.index) {
						this.state.lastIndex = itemHolder.index - 1;
					}
				}
			}
		}
		// Find min and max real visible index
		for (index = 0, itemsLength = this.items.length; index < itemsLength; index += 1) {
			itemHolder = this.items[index];
			if (itemHolder.visible) {
				if (minVisibleIndex === undefined || itemHolder.index < minVisibleIndex) {
					minVisibleIndex = itemHolder.index;
				}
				if (maxVisibleIndex === undefined || itemHolder.index > maxVisibleIndex) {
					maxVisibleIndex = itemHolder.index;
				}
			}
		}
		// If we don't have visible items
		if (minVisibleIndex === undefined && maxVisibleIndex === undefined) {
			// Create fake bottom index
			this.state.lastIndex =
				Math.min(Math.floor(Math.abs(-newPosition_elPadCoord / this.state.itemSize)), itemsCount) || -1;
		}
		// Fill bottom part by new items
		while (
		(this.state.lastIndex < itemsCount - 1)
		&&
		(this.state.lastIndex * this.state.itemSize + newPosition_elPadCoord < this.state.containerSize)
			) {
			this.state.lastIndex += 1;
			this.applyItem(this.state.lastIndex);
		}
		// Fill top part with new items
		if (typeof minVisibleIndex === "number") {
			while (minVisibleIndex > 0 && minVisibleIndex * this.state.itemSize + newPosition_elPadCoord > 0) {
				minVisibleIndex -= 1;
				this.applyItem(minVisibleIndex);
			}
		}
		// Hide invisible items
		this.itemsInvisible.forEach(this.hideItem, this);
	},
	/**
	 * Hide item
	 * @param {Object} itemHolder
	 */
	hideItem: function (itemHolder) {
		propSet(itemHolder.item.style, this.isVertical ? "left" : "top" ,"200%");
	},
	/**
	 * Get height or width of item
	 * @return {Number}
	 */
	getItemSize: function () {
		var size, item;
		item = this.adapter.getItem(0, null);
		this.elWrapper.appendChild(item);
		size = propGet(item.getBoundingClientRect(), this.isVertical ? "height": "width");
		this.elWrapper.removeChild(item);
		return size;
	},
	/**
	 * Set item position in container
	 * @param {Object} itemHolder
	 */
	setItemPosition: function (itemHolder) {
		propSet(itemHolder.item.style, this.isVertical ? "top" : "left", itemHolder.position + "px");
		propSet(itemHolder.item.style, this.isVertical ? "left" : "top" , 0);
	},
	/**
	 * Create item holder if not exists
	 * @param {Number} index
	 * @return {Object}
	 */
	createNewItemHolder: function (index) {
		var item = this.adapter.getItem(index, null), itemHolder;
		this.elWrapper.appendChild(item);
		itemHolder = new ItemHolder(index, item);
		this.items.push(itemHolder);
		return itemHolder;
	},
	/**
	 * Apply item to container
	 * @param {Number} index
	 */
	applyItem: function (index) {
		var itemHolder = this.itemsInvisible.pop() || this.createNewItemHolder(index);
		itemHolder.position = index * this.state.itemSize;
		itemHolder.index = index;
		itemHolder.visible = true;
		this.adapter.getItem(itemHolder.index, itemHolder.item);
		this.setItemPosition(itemHolder);
	},
	/**
	 * Refresh scroll
	 * @param {boolean=true} noDelay
	 */
	refresh: function (noDelay) {
		var delay = noDelay ? 0 : 150;
		clearTimeout(this.refreshID);
		this.refreshID = setTimeout(function () {
			// Stop filling
			window.cancelAnimationFrame(this.requestAnimationFrameID);
			this.state.inAnimation = false;
			// Update state
			this.state.containerPadding =
				propGet(this.elWrapper.getBoundingClientRect(), this.isVertical ? "top" : "left");
			this.state.itemSize = this.getItemSize();
			this.state.containerSize = this.elContainer.offsetHeight;
			// Update wrapper size
			propSet(this.elWrapper.style, this.isVertical ? "height" : "width",
				this.state.itemSize * this.adapter.getCount() + "px");
			// Recalculate items positions
			for (var index = 0, itemLength = this.items.length; index < itemLength; index += 1) {
				this.items[index].position = this.items[index].index * this.state.itemSize;
			}
			// Fill list
			this.fillList(this.state.containerPadding, this.state.containerPadding);
		}.bind(this), delay);
	}
};

	return InfiniteList;
}));
